File: game_plugins/hase_und_igel_new/shared/sc/plugin2018/Field.java
package sc.plugin2018;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamAsAttribute;

@XStreamAlias(value = "field")
public class Field {

  @XStreamAsAttribute
  private int index;
  
  @XStreamAsAttribute
  private FieldType type;
  
  public Field(FieldType type) {
    this(type, -1);
  }
  
  private Field(FieldType type, int index) {
    this.index = index;
    this.type = type;
  }
  
  public int getIndex() {
    return index;
  }

  public void setIndex(int index) {
    this.index = index;
  }

  public FieldType getType() {
    return type;
  }

  public void setType(FieldType type) {
    this.type = type;
  }
  
  @Override
  public Field clone() throws CloneNotSupportedException {
    return new Field(this.type, this.index);
  }
  
}

File: plugin/src/main/kotlin/sc/plugin2024/Field.kt
package sc.plugin2024

import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamAsAttribute
import sc.api.plugins.CubeDirection
import sc.api.plugins.IField

@XStreamAlias("field")
sealed class Field: IField<Field> {
    override val isEmpty
        get() = true
    override fun clone() = this
    
    override fun toString() = javaClass.simpleName
    
    val letter: Char
        get() = javaClass.simpleName.first()
    
    /** Wasserfeld, auf ihm kann sich normal bewegt werden */
    @XStreamAlias("water")
    object WATER : Field() {
        private fun readResolve(): Any = WATER
    }
    
    /** Inselfeld, es kann nicht überwunden werden und kein Spieler kann darauf stehen */
    @XStreamAlias("island")
    object BLOCKED: Field() {
        private fun readResolve(): Any = BLOCKED
        override val isEmpty
            get() = false
    }

    /** Passagierfeld mit Anleger */
    @XStreamAlias("passenger")
    data class PASSENGER(@XStreamAsAttribute val direction: CubeDirection = CubeDirection.values().random(), @XStreamAsAttribute var passenger: Int = 1): Field() {
        override val isEmpty
            get() = false
        override fun clone() = PASSENGER(direction, passenger)
        override fun toString() = "PASSENGER${direction.ordinal}${passenger}"
    }

    /** Ein Zielfeld */
    @XStreamAlias("goal")
    object GOAL: Field() {
        private fun readResolve(): Any = GOAL
    }
    
    /** Ein Sandbankfeld */
    @XStreamAlias("sandbank")
    object SANDBANK: Field() {
        private fun readResolve(): Any = SANDBANK
    }
}